kaitai/kaitaistream.cpp

```diff
@@ -251,20 +251,24 @@ uint64_t kaitai::kstream::read_u8le() {
 
 float kaitai::kstream::read_f4be() {
     uint32_t t;
+    float out;
     m_io->read(reinterpret_cast<char *>(&t), 4);
 #if __BYTE_ORDER == __LITTLE_ENDIAN
     t = bswap_32(t);
 #endif
-    return reinterpret_cast<float&>(t);
+    type_pun(out, t);
+    return out;
```

Comment:
Can you explain why `type_pun` was actually needed, and what's wrong with `reinterpret_cast` solution? `type_pun` implementation seems to do a call to `std::memcpy` — I'm not sure what's the purpose of that and if that ever gets optimized out, as we don't really need any memory copying here, it's just a matter of reinterpreting existing memory contents (likely in RAX, not even on stack).

Reply:
`type_pun` is just a utility. `reinterpret_cast` in this case [violates strict aliasing rules](https://github.com/CppCon/CppCon2017/raw/master/Presentations/Type%20Punning%20In%20C%2B%2B17%20-%20Avoiding%20Pun-defined%20Behavior/Type%20Punning%20In%20C%2B%2B17%20-%20Avoiding%20Pun-defined%20Behavior%20-%20Scott%20Schurr%20-%20CppCon%202017.pdf) (Page 70). The call to `std::memcpy` is optimized, doesn't violate aliasing, and the code produced is the same as if using `reinterpret_cast`.